File: src/Алгоритмы и структуры данных/Lab3/task1.py
import matplotlib.pyplot as plt
import networkx as nx

class Node:
    def __init__(self, key):
        self.left = None
        self.right = None
        self.val = key

def insert(root, key):
    if root is None:
        return Node(key)
    else:
        if key < root.val:
            root.left = insert(root.left, key)
        else:
            root.right = insert(root.right, key)
    return root

def plot_tree(root):
    tree = nx.DiGraph()
    pos = {}
    
    def add_edges(node, x=0, y=0, layer=1):
        if node:
            pos[node.val] = (x, y)
            if node.left:
                tree.add_edge(node.val, node.left.val)
                add_edges(node.left, x - 1/layer, y - 1, layer + 1)
            if node.right:
                tree.add_edge(node.val, node.right.val)
                add_edges(node.right, x + 1/layer, y - 1, layer + 1)
    
    add_edges(root)
    plt.figure(figsize=(8, 5))
    nx.draw(tree, pos, with_labels=True, arrows=False, node_size=2000, node_color='lightblue')
    plt.title("Бинарное дерево поиска (BST)")
    plt.show()


keys = [50, 20, 70, 10, 30, 60, 90]
root = None
for key in keys:
    root = insert(root, key)
plot_tree(root)

File: src/Алгоритмы и структуры данных/Lab3/task1.java
import java.util.*;

class Node {
    int val;
    Node left, right;
    
    public Node(int val) {
        this.val = val;
        left = right = null;
    }
}

class BST {
    Node root;
    
    void insert(int key) {
        root = insertRec(root, key);
    }
    
    Node insertRec(Node root, int key) {
        if (root == null) {
            root = new Node(key);
            return root;
        }
        if (key < root.val)
            root.left = insertRec(root.left, key);
        else if (key > root.val)
            root.right = insertRec(root.right, key);
        return root;
    }
    
    void inorder() {
        inorderRec(root);
    }
    
    void inorderRec(Node root) {
        if (root != null) {
            inorderRec(root.left);
            System.out.print(root.val + " ");
            inorderRec(root.right);
        }
    }
}

public class task1 {
    public static void main(String[] args) {
        BST tree = new BST();
        int[] keys = {50, 20, 70, 10, 30, 60, 90};
        for (int key : keys) {
            tree.insert(key);
        }
        
        System.out.println("Inorder обход BST:");

        tree.insert(25);
        tree.inorder();

        
    }
}